lib/puppet/provider/vcsrepo/git.rb

```diff
@@ -353,6 +354,16 @@ def update_owner_and_excludes
     end
   end
 
+  def remote_source
+    source = @resource.value(:source)
+    if @resource.value(:basic_auth_username) && @resource.value(:basic_auth_password)
+      uri = URI(@resource.value(:source))
+      uri.userinfo = "#{@resource.value(:basic_auth_username)}:#{@resource.value(:basic_auth_password)}"
+      source = uri.to_s
+    end
+    source
+  end
```

Comment:
Suggested rewrite - allows for empty passwords, avoids the extra `source` variable:
``` ruby
def remote_source
  if @resource.value(:basic_auth_username)
    uri = URI(@resource.value(:source))
    uri.userinfo = "#{@resource.value(:basic_auth_username)}:#{@resource.value(:basic_auth_password)}"
    uri.to_s
  else
    @resource.value(:source)
  end
end